llvm/lib/CheerpUtils/PartialExecuter.cpp

```diff
@@ -0,0 +1,1393 @@
+//===-- Cheerp/PartialExecuter.h - Analyze functions CFG to mark unreachable BBs --===//
+//
+//                     Cheerp: The C++ compiler for the Web
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+// Copyright 2021-2022 Leaning Technologies
+//
+//===----------------------------------------------------------------------===//
+
+#define DEBUG_TYPE "PartialExecuter"
+
+#include "../ExecutionEngine/Interpreter/Interpreter.h"
+#include "llvm/ADT/SCCIterator.h"
+#include "llvm/Cheerp/DeterministicUnorderedSet.h"
+#include "llvm/Cheerp/PartialExecuter.h"
+#include "llvm/ExecutionEngine/GenericValue.h"
+#include "llvm/IR/Operator.h"
+#include "llvm/IR/InstIterator.h"
+#include "llvm/InitializePasses.h"
+#include <map>
+#include <unordered_map>
+#include <vector>
+
+using namespace llvm;
+
+typedef cheerp::DeterministicUnorderedSet<const BasicBlock *, cheerp::RestrictionsLifted::NoErasure> DeterministicBBSet;
+
+namespace cheerp {
+
+char PartialExecuter::ID = 0;
+
+StringRef PartialExecuter::getPassName() const
+{
+	return "PartialExecuter";
+}
+
+PartialExecuter::PartialExecuter()
+	: llvm::ModulePass(ID)
+{
+}
+
+void PartialExecuter::getAnalysisUsage(AnalysisUsage& AU) const
+{
+}
+
+//Compute whether a given Value is a constant & it's value can be executed beforehand
+static bool isValueComputedConstant(const llvm::Value* V)
+{
+	// TODO(carlo): this can easily be memoized since there is no state
+	if (!isa<Constant>(V))
+		return false;
+	if (const GlobalVariable* GVar = dyn_cast<GlobalVariable>(V))
+	{
+		if (!GVar->hasInitializer() || GVar->isExternallyInitialized())
+			return false;
+		if (!GVar->hasInternalLinkage())
+			return false;
+	}
+	if (const ConstantExpr* CE = dyn_cast<ConstantExpr>(V))
+	{
+		for (auto& op : CE->operands())
+		{
+			if (!isValueComputedConstant(op))
+				return false;
+		}
+	}
+	return true;
+}
+
+class FunctionData;
+class ModuleData;
+
+class PartialInterpreter : public llvm::Interpreter {
+	friend FunctionData;
+	friend ModuleData;
+	const llvm::BasicBlock* fromBB{nullptr};	
+	enum BitMask : uint32_t
```

Comment:
I have doubts about the behavior of 64-bit integers, are those taken into account?

Reply:
I may advise to use a shorter value such as ALL = 0xff, then. The idea would be that the current mask _might_ suggest this to be 32-bit specialized, while it's not.

Reply:
Perfect, done